src/esp/gfx/Renderer.cpp

```diff
@@ -74,6 +70,52 @@ struct Renderer::Impl {
   void draw(RenderCamera& camera, MagnumDrawableGroup& drawables) {
     renderEnter();
     camera.getMagnumCamera().setViewport(framebufferSize_);
+
+    /* Inverted projection matrix to unproject the depth value and chop the
+       near plane off. We don't care about X/Y there and the corresponding
+       parts of the matrix are zero as well so take just the lower-right part
+       of it (denoted a, b, c, d).
+
+        x 0 0 0
+        0 y 0 0
+        0 0 a b
+        0 0 c d
+
+       Doing an inverse of just the bottom right block is enough as well -- see
+       https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices for
+       a proof.
+
+       Taking a 2-component vector with the first component being Z and second
+       1, the final calculation of unprojected Z is then
+
+        | a b |   | z |   | az + b |
+        | c d | * | 1 | = | cz + d |
+
```

Comment:
I think it may be possible to solve the perspective projection for z.

```
Given:

projectionMatrix[2].z = A
projectionMatrix[3].z = B
projectionMatrix[2].w = -1
projectionMatrix[3].w = 0

zperspective = | A  B |   | z |   | Az + B | = (Az + B)/-z
               | -1 0 | * | 1 | = |  -z    |

zperspective = 2.0 * depth - 1.0

Plugging into above:

2.0 * depth - 1.0 = (Az + B)/-z

Solving for depth:

z = -B/(2.0 * depth + A - 1.0)

Inverting (to get positive values):

z = B/(2.0 * depth + A - 1.0)

If addition is used for 2.0 * depth, only 2 additions and 1 divide are required.
```

Reply:
Thank you!

Will apply this in a separate PR together with setting up C++-side testing for this -- wanted to measure the perf difference for this :)

Reply:
Update after benchmarking -- this makes the calculation slightly faster (thanks again), however the most time (>75%) is consumed by the `if()` that flips values at the far plane to `0.0f` :/ In fact, even doing the dumbest full 4x4 multiplication and extracting just Z afterwards is a minor perf penalty compared to costs of the `if`.

It gets a bit faster when done in a separate loop, however that makes it ~~no longer possible~~ much harder to do the unprojection in-place.

Cc: @erikwijmans

Reply:
Can you try a ternary instead of an `if`, i.e. `ptr[i] = z == 1.0 ? 0.0 : <unprojected>`?  Jumps are muuuccch than conditional moves and the compute cost of getting the unprojected value is so trivial that computing it more often than needed shouldn't matter :)

Reply:
No change, as expected -- I assume the optimizer would try that as a first thing.

Problem is that the ifs prevent code from being batched & vectorized. If I move the condition to a separate loop it gets three times faster, but still about 2x slower than a dumb 4x4 multiplication without the branch would be.

Anyway, will submit this together with the GPU version tomorrow. Maybe you figure out an improvement again :)

Reply:
Ah, yeah.  The ternary won't help with the CPU version, but I believe it'll help with the OpenGL version (it 100% would with the CUDA version, if's are terrible on CUDA).